# CMPS 240

# Assignment 2

**Due March 3, 2019 11:59**

## Problem 1

Consider the following constraint satisfaction problem. A graph has nodes of the following types:
- Triangle
- Circle
- Square
- Hexagon
- Pentagon

Each node has a domain of {1, 2, ..., 9}.

Each node type as the following constraints on its value:
- Triangle - The leftmost digit of the product of all of its neightbors
- Square - The rightmost digit of of the product of all its neighbors
- Hexagon - The leftmost digit of the sum of all its neighbors
- Pentagon - The rightmost digit of the sum of all its neighbors
- Circle - No contraints

Complete the function defined below:

```python
def solve_csp(nodes, arcs, max_steps):
    """
    This function solves the csp using the MinConflicts Search
    Algorithm.

    INPUTS:
    nodes:      a list of letters that indicates what type of node it is,
                the index of the node in the list indicates its id
                letters = {C, T, S, P, H}
    arcs:       a list of tuples that contains two numbers, indicating the 
                IDS of the nodes the arc connects. 
    max_steps:  max number of steps to make before giving up

    RETURNS: a list of values for the soltiion to the CSP where the 
             index of the value correxponds the the value for that
             given node.
    """
    node_values = []

    return node_values
```

As a reminder here is the pseudo code for the Min-Conflicts search algorithm:

![minconflicts](https://docs.google.com/drawings/d/e/2PACX-1vTIdOyAKDEoK6evNWQBkx9X5kl2I7GLaUkE9TdFDRqyyNFiHeFDrA-Sm7sLob2wMSzoBk_cliRhs8PY/pub?w=927&amp;h=474)

**Notes:**

- It's possible that you won't converge to a solution in a single run. Try a few runs to see if you get to a solution.
- The example is to show you what a problem looks like, I will test/grade your program on different examples

## Question 1

In [0]:
import random as random
import numpy as numpy

In [0]:
   """
    This function solves the csp using the MinConflicts Search
    Algorithm.

    INPUTS:
    nodes:      a list of letters that indicates what type of node it is,
                the index of the node in the list indicates its id
                letters = {C, T, S, P, H}. From input
    node_values: current assignments
    adjmat     : adjacency matrix
    iternode   : this iteration node
    
    RETURNS: The new value to be assigned to the node under consideration
   """
   def conflicts_util(nodes, node_values, adjmat, iternode):
      listindex = []
      listnum = []
      for i in range(1,10):
        node_values[iternode]=i
        listindex.append(i);
        listnum.append(len(conflicts_count(nodes,node_values,adjmat)))
      min_num = min(listnum);
      minlist=[i for i, j in zip(listindex, listnum) if j == min_num]
      randomindex = random.choice(minlist)
      return randomindex
  
   def conflicts_count(nodes, node_values, adjmat):    
      """
      This function conflicts_count looks at the constraints and updates the val
      for the nodes
      INPUTS:
      nodes      : a list of letters that indicates what type of node it is,
                   the index of the node in the list indicates its id
                letters = {C, T, S, P, H}. From input
      node_values: current assignments
      adjmat     : adjacency matrix
      
      RETURNS    : new set of conflict list after new assignments 
      """
      update_list=[]
      for node in nodes:
        if node == 'T':
          #Triangle - The leftmost digit of the product of all of its neightbors
          i = nodes.index(node)
          list2=[node_values[x] for x in adjmat[i]] 
          digit = numpy.prod(list2)
          digit = int(str(digit)[0])
          update_list.insert(i,digit)
          
        if node == 'S':
          #Square - The rightmost digit of of the product of all its neighbors
          i = nodes.index(node)
          list2=[node_values[x] for x in adjmat[i]] 
          digit = numpy.prod(list2)
          digit = digit %10
          if(digit != 0):
            update_list.insert(i,digit)
          else:
            continue
            
        if node == 'H':
          #Hexagon - The leftmost digit of the sum of all its neighbors
          i = nodes.index(node)
          list2=[node_values[x] for x in adjmat[i]] 
          digit = sum(list2)
          digit = int(str(digit)[0])
          if(digit != 0):
            update_list.insert(i,digit)
          else:
            continue
            
        if node == 'P':
          #Pentagon - The rightmost digit of the sum of all its neighbors
          i = nodes.index(node)
          list2=[node_values[x] for x in adjmat[i]] 
          rdigit = sum(list2)
          rdigit = rdigit %10
          update_list.insert(i,rdigit)
          
        if node == 'C':
          #Circle - No contraints
          i = nodes.index(node)
          val = node_values[i]
          update_list.insert(i,val)
      
      temp = []
      for i in range(0,len(node_values)):
        for j in range(0,len(update_list)):
          if(i==j and node_values[i] != update_list[j] ):
            temp.append(i)
      #print("temp is ",temp)
      return (temp)
  
  
def solve_csp(nodes, arcs, max_steps):
    """
    This function solves the csp using the MinConflicts Search
    Algorithm.

    INPUTS:
    nodes:      a list of letters that indicates what type of node it is,
                the index of the node in the list indicates its id
                letters = {C, T, S, P, H}
    arcs:       a list of tuples that contains two numbers, indicating the 
                IDS of the nodes the arc connects. 
    max_steps:  max number of steps to make before giving up

    RETURNS: a list of values for the soltiion to the CSP where the 
             index of the value correxponds the the value for that
             given node.
    """
    node_values = []
    n = len(nodes)
    #create empty 2D matrix
    adjmat = [[] for _ in range(n)]
    
    #create adj matrix
    for node in nodes:
      i = nodes.index(node)
      for arc in arcs:
        if(i in arc):
          adjmat[i].append(arc[0])
          adjmat[i].append(arc[1])
       
      adjmat[i]=set(adjmat[i])
      adjmat[i].remove(i)
    
    #check if adacency matrix was populated
    print("adjmat is ", adjmat)
        
    #Fill up with initial values
    for i in range(n):
      node_values.insert(i, random.randint(1, 9));
    print("node values ",node_values)
    
    #inital conflicts
    conf = conflicts_count(nodes,node_values,adjmat)

    #min conflicts algorithm
    for i in range(max_steps):
      if len(conf) == 0:
        #ANSWER FOUND!! 
        print(node_values)
        return node_values
      else:
        iternode = random.choice(conf)
        newval = conflicts_util(nodes, node_values, adjmat, iternode)
        node_values[iternode]=newval
        conf = conflicts_count(nodes, node_values, adjmat)
            
    print("Reached max_steps i.e. ",max_steps,"Converge failed!")
    return node_values


In [3]:
# Here is an exmaple input to test your code on. It is solveable.
nodes = 'CHTPS'
arcs = [(0,1), (0,2), (1,2), (1,3), (1,4), (2,3), (2,4)]
max_steps = 1000
solve_csp(nodes, arcs, max_steps)

# E.g. pass test case: input
# adjmat is  [{1, 2}, {0, 2, 3, 4}, {0, 1, 3, 4}, {1, 2}, {1, 2}]
# node values  [4, 9, 1, 8, 5]
# Output:
# [4, 1, 2, 3, 2]
#


adjmat is  [{1, 2}, {0, 2, 3, 4}, {0, 1, 3, 4}, {1, 2}, {1, 2}]
node values  [9, 3, 6, 8, 2]
Reached max_steps i.e.  1000 Converge failed!


[9, 2, 6, 8, 2]

## Problem 2

Solve the following  MDP using both value iteration and policy iteration, you can do this by hand or programmatically, but you need to show your work in either case. 

There is a self-driving taxi that takes from place to place. Its goal is to make the most money possible and it makes the most money in a particular town, MoneyTown. The taxi has a tendency to take routes that take it to different towns and it costs money for the taxi to drive from place to place.  

There are three states that the taxi can be in: 'In MoneyTown', 'MoneyTown Suburbs', and 'Outside MoneyTown'. There are two actions that the taxi can take in each state: drive and wait. Driving costs \$10. When the taxi is in money town it makes \$30, in MoneyTown Suburbs and Outside MoneyTown it only makes \$10. The reward for the taxi is:

(money made - cost) 

For example if the taxi is driving around in MoneyTown, the reward is \$30-\$10=\$20.

If the taxi is in MoneyTown and drives, then it is still MoneyTown in the next period with probability .9, and in the MoneyTown Suburbs in the next period with probability .1. If it is MoneyTown and does not drive, these probabilities are .7 and .3, respectively. If it is in the MoneyTown Suburbs and drives, then with probability .3 it is in MoneyTown in the next period, with probability .6 it is still in MoneyTown Suburbs in the next period, and with probability .1 it is in Outside MoneyTown in the next period. If it is in MoneyTown Suburbs and does not drive, then with probability 1 it is Outside MoneyTown next period. Finally, if it is in Outside MoneyTown and drives, then in the next period it is in MoneyTown with probability .6, and at the OutSide MoneyTown with probability .4. If it does not drive, then with probability 1 it is at Outside MoneyTown in the next period. 

1. Draw the MDP graphically
  - A good way to do this is through [Google Drawings](https://docs.google.com/drawings)
  - When you're done you can embed it in the jupyter notebook using markdown syntax
  - \!\[alt-text\]\(url\)
  - To get the URL for your image in Google Draw goto File->Publish to the web...->Embed and copy the src portion of the html tag
  
2. Using a discount factor of .8, solve the MDP using value iteration (until the values have become reasonably stable). You should start with the values set to zero. You should show both the optimal policy and the optimal values.
3. Using a discount factor of .8, solve the MDP using policy iteration (until you have complete convergence). You should start with the policy that never drives. Again, you should show both the optimal policy and the optimal values (and of course they should be the same as in 2...).
4. Change the MDP in three different ways: by changing the discount factor, changing the transition probabilities for a single action from a single state, and by changing a reward for a single action at a single state. Each of these changes should be performed separately starting at the original MDP, resulting in three new MDPs (which you do not have to draw), each of which is different from the original MDP in a single way. In each case, the change should be so that the optimal policy changes, and you should state what the optimal policy becomes and give a short intuitive argument for this.


**If you solve the problem programmatically, put your code in here. If you solve it by hand include your work here as well. You can add cells as you feel the need.**

## Question 2.1

In [0]:
from IPython.display import HTML
HTML('<img src="https://docs.google.com/drawings/d/e/2PACX-1vSV1qVR1nr-l45prl_fYbX8d2Z7M-GDvu5jiniNBnj8xxylRL-WfOyXOetBCq0mkfzZ450dCLQtvBdm/pub?w=960&amp;h=720">')

##Question 2.3

In [4]:
#given
gamma = 0.8 #discount factor
#0 intown 1 suburb 2 outoftown
states = [0,1,2]
#print(len(states))
#0 drive 1 wait
actions = [0,1]
#print(len(actions))
Trans = numpy.zeros((3, 2, 3))
Reward = numpy.zeros((3, 2, 3))
V = [0, 0 ,0]     #expected utility
Qopt = 0
#---------
Trans[0,0,0] = 0.9
Trans[0,0,1] = 0.1
Trans[0,1,0] = 0.7
Trans[0,1,1] = 0.3
#---------
Trans[1,0,0] = 0.3
Trans[1,0,1] = 0.6
Trans[1,0,2] = 0.1
Trans[1,1,2] = 1
#---------
Trans[2,0,2] = 0.4
Trans[2,0,0] = 0.6
Trans[2,1,2] = 1
#---------
Reward[0,0,0] = 20
Reward[0,0,1] = 0
Reward[0,1,0] = 30
Reward[0,1,1] = 10
#---------
Reward[1,0,0] = 20
Reward[1,0,1] = 0
Reward[1,0,2] = 0
Reward[1,1,2] = 10
#---------
Reward[2,0,2] = 0
Reward[2,0,0] = 20
Reward[2,1,2] = 10
#---------
pie = [0, 0, 0]   #random policy - always drive
iteration = 0

def policy_evaluation():
  for state in states:
    Vpie = 0
    for state1 in states:
      Vpie += Trans[state, pie[state], state1] * (Reward[state, pie[state], state1] + (gamma * V[state1] ))
    V[state] = Vpie
  return V

#policy iteration
unchanged = True
while unchanged:  #do
  unchanged = False
  V = policy_evaluation();
  #print(V)
  #improvement phase 
  for state in states:
    Qopt = V[state]
    for action in actions:
      Qs = 0
      for state1 in states:
        Qs += Trans[state, action, state1] * (Reward[state, action, state1] + (gamma * V[state1] ))
        if (Qs > Qopt):
          pie[state] = action
          Qopt = Qs
          unchanged = True
        

#result
print (pie)
print("By policy iteration Optimal policy is [",end="")
for action in pie:
  if action == 0:
    print("drive ",end="")
  else:
    print("wait ",end="")
print("] with Optimal Values are = ", V)


[1, 1, 0]
By policy iteration Optimal policy is [wait wait drive ] with Optimal Values are =  [97.8361669242658, 79.36630602782068, 86.70788253477585]


## **Question 2.2**

In [5]:
#value iteration
Uopt = [0, 0 ,0]     #expected utility 
epsilon = 0.0001     # error allowed
iteration = 0
def expected_reward(state, action):
  result = 0
  for dest in states:
    result += Reward[state, action, dest] * Trans[state, action, dest]
  return result

while True:

  delta = 0
  #outdate the optimal into prev_U
  prev_U = Uopt.copy()
  for state in states:
    Vmax = 0
    state1max = 0
    actionmax = 0
    #find expected value of each action
    for action in actions:
      V = 0
      for state1 in states:
        V += Trans[state, action, state1] * prev_U[state1]
        if (V > Vmax):
          Vmax = V
          actionmax = action

    Uopt[state] = expected_reward(state, actionmax) + (gamma * Vmax)
    
    delta = max(delta, abs(Uopt[state] - prev_U[state])) 
  #exit if u find change to be too less  
  if (delta < epsilon * (1 - gamma) / gamma):
    break

print("Optimal value is ", Uopt)

#get policy from Uopt or utility function
valiterpie = [0,0,0]
for state in states:
  UofA = [0,0]
  for action in actions:
    for state1 in states:
      UofA[action] += Trans[state,action,state1] * (Reward[state,action,state1] + (gamma * Uopt[state1]))
  valiterpie[state] = numpy.argmax(UofA)
  
print("Optimal policy is ", valiterpie)

#result
print("By value iteration Optimal policy is [",end="")
for i in valiterpie:
  if i == 0:
    print("drive ",end="")
  else:
    print("wait ",end="")
print("]")

Optimal value is  [84.86965778887983, 72.04400648627463, 77.5550285303628]
Optimal policy is  [1, 1, 0]
By value iteration Optimal policy is [wait wait drive ]


**Comment : **The policy from value iteration and policy iteration are seen to be the same i.e [wait wait drive ]

## Question 2.4

## Question 2.4.1 - Discount factor, Gamma changed

In [4]:
#given
gamma = 0.99 #discount factor
#0 intown 1 suburb 2 outoftown
states = [0,1,2]
#print(len(states))
#0 drive 1 wait
actions = [0,1]
#print(len(actions))
Trans = numpy.zeros((3, 2, 3))
Reward = numpy.zeros((3, 2, 3))
V = [0, 0 ,0]     #expected utility
Qopt = 0
#---------
Trans[0,0,0] = 0.9
Trans[0,0,1] = 0.1
Trans[0,1,0] = 0.7
Trans[0,1,1] = 0.3
#---------
Trans[1,0,0] = 0.3
Trans[1,0,1] = 0.6
Trans[1,0,2] = 0.1
Trans[1,1,2] = 1
#---------
Trans[2,0,2] = 0.4
Trans[2,0,0] = 0.6
Trans[2,1,2] = 1
#---------
Reward[0,0,0] = 20
Reward[0,0,1] = 0
Reward[0,1,0] = 30
Reward[0,1,1] = 10
#---------
Reward[1,0,0] = 20
Reward[1,0,1] = 0
Reward[1,0,2] = 0
Reward[1,1,2] = 10
#---------
Reward[2,0,2] = 0
Reward[2,0,0] = 20
Reward[2,1,2] = 10
#---------
pie = [0, 0, 0]   #random policy - always drive
iteration = 0

def policy_evaluation():
  for state in states:
    Vpie = 0
    for state1 in states:
      Vpie += Trans[state, pie[state], state1] * (Reward[state, pie[state], state1] + (gamma * V[state1] ))
    V[state] = Vpie
  return V

#policy iteration
unchanged = True
while unchanged:  #do
  unchanged = False
  V = policy_evaluation();
  #print(V)
  #improvement phase 
  for state in states:
    Qopt = V[state]
    for action in actions:
      Qs = 0
      for state1 in states:
        Qs += Trans[state, action, state1] * (Reward[state, action, state1] + (gamma * V[state1] ))
        if (Qs > Qopt):
          pie[state] = action
          Qopt = Qs
          unchanged = True
        

#result
print (pie)
print("By policy iteration Optimal policy is [",end="")
for action in pie:
  if action == 0:
    print("drive ",end="")
  else:
    print("wait ",end="")
print("] with Optimal Values are = ", V)


[1, 1, 0]
By policy iteration Optimal policy is [wait wait drive ] with Optimal Values are =  [1839.4197980727738, 1820.5450437991308, 1828.8333775748797]


**Policy Iteration output**

Output for Gamma = 0.99

[1, 1, 0]
By policy iteration Optimal policy is [wait wait drive ] with Optimal Values are =  [1839.4197980727738, 1820.5450437991308, 1828.8333775748797]

Output for Gamma = 0.5

[1, 1, 0]
By policy iteration Optimal policy is [wait wait drive ] with Optimal Values are =  [42.81407035175878, 25.527638190954768, 31.055276381909536]

Output for Gamma = 0.1

[1, 1, 0]
By policy iteration Optimal policy is [wait wait drive ] with Optimal Values are =  [26.174631982254486, 11.41359144988909, 14.135914498890903]

Output for Gamma = 0.01 

[1, 1, 0]
By policy iteration Optimal policy is [wait wait drive ] with Optimal Values are =  [24.19976416841338, 10.121939744829422, 12.193974482942249]



**Intuitive Argument:** This result makes sense because, this is an infinite MDP and optimal policy is obtained not based on a final goal. So collecting the rewards immediately or later does not matter which is the soul purpose of the discount factor

## **Question 2.4.2 Transition probability changed** Trans[1,0,0] = 0.98 (balance other probabilties)

In [18]:
#given
gamma = 0.8 #discount factor
#0 intown 1 suburb 2 outoftown
states = [0,1,2]
#print(len(states))
#0 drive 1 wait
actions = [0,1]
#print(len(actions))
Trans = numpy.zeros((3, 2, 3))
Reward = numpy.zeros((3, 2, 3))
V = [0, 0 ,0]     #expected utility
Qopt = 0
#---------
Trans[0,0,0] = 0.9
Trans[0,0,1] = 0.1
Trans[0,1,0] = 0.7
Trans[0,1,1] = 0.3
#---------
Trans[1,0,0] = 0.98
Trans[1,0,1] = 0.01
Trans[1,0,2] = 0.01
Trans[1,1,2] = 1
#---------
Trans[2,0,2] = 0.4
Trans[2,0,0] = 0.6
Trans[2,1,2] = 1
#---------
Reward[0,0,0] = 20
Reward[0,0,1] = 0
Reward[0,1,0] = 30
Reward[0,1,1] = 10
#---------
Reward[1,0,0] = 20
Reward[1,0,1] = 0
Reward[1,0,2] = 0
Reward[1,1,2] = 10
#---------
Reward[2,0,2] = 0
Reward[2,0,0] = 20
Reward[2,1,2] = 10
#---------
pie = [0, 0, 0]   #random policy - always drive
iteration = 0

def policy_evaluation():
  for state in states:
    Vpie = 0
    for state1 in states:
      Vpie += Trans[state, pie[state], state1] * (Reward[state, pie[state], state1] + (gamma * V[state1] ))
    V[state] = Vpie
  return V

#policy iteration
unchanged = True
while unchanged:  #do
  unchanged = False
  V = policy_evaluation();
  #print(V)
  #improvement phase 
  for state in states:
    Qopt = V[state]
    for action in actions:
      Qs = 0
      for state1 in states:
        Qs += Trans[state, action, state1] * (Reward[state, action, state1] + (gamma * V[state1] ))
        if (Qs > Qopt):
          pie[state] = action
          Qopt = Qs
          unchanged = True
        

#result
print (pie)
print("By policy iteration Optimal policy is [",end="")
for action in pie:
  if action == 0:
    print("drive ",end="")
  else:
    print("wait ",end="")
print("] with Optimal Values are = ", V)

[1, 0, 0]
By policy iteration Optimal policy is [wait drive drive ] with Optimal Values are =  [115.5868902439024, 111.90929878048777, 99.23780487804875]



[1, 0, 0]

By policy iteration Optimal policy is [wait drive drive ] with Optimal Values are =  [115.5868902439024, 111.90929878048777, 99.23780487804875]


**Intuitive Argument:** This makes sense because we give so much high probability(0.98) that moving from state 1 to 0 will happen that the agent decides to take the drive and the bet for a high reward safely.

## **Question 2.4.3 Rewards changed Reward[1,0,2] = 80**

In [17]:
#given
gamma = 0.9988 #discount factor
#0 intown 1 suburb 2 outoftown
states = [0,1,2]
#print(len(states))
#0 drive 1 wait
actions = [0,1]
#print(len(actions))
Trans = numpy.zeros((3, 2, 3))
Reward = numpy.zeros((3, 2, 3))
V = [0, 0 ,0]     #expected utility
Qopt = 0
#---------
Trans[0,0,0] = 0.9
Trans[0,0,1] = 0.1
Trans[0,1,0] = 0.7
Trans[0,1,1] = 0.3
#---------
Trans[1,0,0] = 0.3
Trans[1,0,1] = 0.6
Trans[1,0,2] = 0.1
Trans[1,1,2] = 1
#---------
Trans[2,0,2] = 0.4
Trans[2,0,0] = 0.6
Trans[2,1,2] = 1
#---------
Reward[0,0,0] = 20
Reward[0,0,1] = 0
Reward[0,1,0] = 30
Reward[0,1,1] = 10
#---------
Reward[1,0,0] = 20
Reward[1,0,1] = 0
Reward[1,0,2] = 80
Reward[1,1,2] = 10
#---------
Reward[2,0,2] = 0
Reward[2,0,0] = 20
Reward[2,1,2] = 10
#---------
pie = [0, 0, 0]   #random policy - always drive
iteration = 0

def policy_evaluation():
  for state in states:
    Vpie = 0
    for state1 in states:
      Vpie += Trans[state, pie[state], state1] * (Reward[state, pie[state], state1] + (gamma * V[state1] ))
    V[state] = Vpie
  return V

#policy iteration
unchanged = True
while unchanged:  #do
  unchanged = False
  V = policy_evaluation();
  #print(V)
  #improvement phase 
  for state in states:
    Qopt = V[state]
    for action in actions:
      Qs = 0
      for state1 in states:
        Qs += Trans[state, action, state1] * (Reward[state, action, state1] + (gamma * V[state1] ))
        if (Qs > Qopt):
          pie[state] = action
          Qopt = Qs
          unchanged = True
        

#result
print (pie)
print("By policy iteration Optimal policy is [",end="")
for action in pie:
  if action == 0:
    print("drive ",end="")
  else:
    print("wait ",end="")
print("] with Optimal Values are = ", V)

[1, 0, 0]
By policy iteration Optimal policy is [wait drive drive ] with Optimal Values are =  [16007.196480298095, 15991.206077736222, 15995.191691168802]


Reward[1,0,2] = 0

[1, 1, 0]

By policy iteration Optimal policy is** [wait wait drive ]** with Optimal Values are =  [97.8361669242658, 79.36630602782068, 86.70788253477585]

Reward[1,0,2] = 80

[1, 0, 0]

By policy iteration Optimal policy is **[wait drive drive ]** with Optimal Values are =  [102.62686567164177, 88.14925373134328, 90.08955223880596]

**Intuitive Argument:** This makes sense because we claim to give a better reward for waiting while on the move from 1 to 2. So the optimal state at state 1 changed from wait to drive to 2 so that the 80 is claimed
